Feature/timelines custom fields display 3884

nkenneweg opened the pull request:
Implements [3884](https://www.openproject.org/work_packages/3884) and [3114](https://www.openproject.org/work_packages/3114)

nkenneweg committed:
bug: featuresteps: custom field set,wp field check

nkenneweg committed:
timelines: custom field filters.

for now only for list and boolean custom fields.

nkenneweg committed:
intendation changed.

nkenneweg committed:
pass timelines_id for local custom fields filter.

nkenneweg committed:
integrative timelines custom field filter tests.

also adds these steps:
- the custom field <name> is enabled for the project <name>
- when i filter for work packages with the boolean custom field <name> set to <value>
- when i filter for work packages with the list custom field <name> set to <value>

nkenneweg committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_filter_custom_fields_3701

nkenneweg committed:
changelog.

nkenneweg committed:
use only custom fields for this project.

nkenneweg committed:
fix tests. user needs view work package permission.

nkenneweg committed:
add custom field values to api.

nkenneweg committed:
display custom fields in timeline.

nkenneweg committed:
api test.

nkenneweg committed:
use project-specific custom field api.

nkenneweg committed:
integrative tests for display.

nkenneweg commented:
This depends on #834 due to new test-steps.

apfelfabrik committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_filter_custom_fields_3701

Conflicts:
	doc/CHANGELOG.md

apfelfabrik commented:
Rebased onto feature/timelines_filter_custom_fields_3701, since it diverged.

apfelfabrik commented:
Four tests keep failing on me, the branch this depends on also has a remaining issue, so not merging this yet.

apfelfabrik committed:
Fixes filter for (none) in custom fields.

The id -1 and the empty string both need to be added as values in the
(none)-case, since (none) has to both match work packages with custom
fields with an empty value and work packages with no custom field.

apfelfabrik committed:
Merge branch 'feature/timelines_filter_custom_fields_3701' into feature/timelines_custom_fields_display_3884

apfelfabrik committed:
Fixes specs.

Adds empty custom values array to the generated structs, as those are
now appended on the server, even when there are no custom fields. This
also fixes minor whitespace issues.

apfelfabrik committed:
Fixes a bug w/ (none) combined with another selection.

Of course, (none) can be selected in a list of several other options, in
which case we need the empty string fix, too.

apfelfabrik committed:
Merge branch 'feature/timelines_filter_custom_fields_3701' into feature/timelines_custom_fields_display_3884

nkenneweg committed:
fix customfieldvalue can be a string not an array.

"-1".indexOf("-1") is true which leads to value.push failing.
value can be a string e.g. for filters for boolean.

nkenneweg committed:
fix. rabl nodes are added to every element.

if we do node("a") {"b"} in one of the pes, every
pe gets a node a with value b, even if we do not add it again.
creating a Hash and mapping the values correctly is also
a lot cleaner.

nkenneweg committed:
added a small test.

nkenneweg committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_custom_fields_display_3884

Conflicts:
	doc/CHANGELOG.md

nkenneweg committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_custom_fields_display_3884

nkenneweg committed:
changelog.
